File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        count = {}
        left = 0
        max_count = 0
        max_length = 0
        
        for right in range(len(s)):
            # Increment the count of the current character
            count[s[right]] = count.get(s[right], 0) + 1
            
            # Update the max_count to be the highest frequency character in the current window
            max_count = max(max_count, count[s[right]])
            
            # If the number of characters to replace exceeds k, shrink the window
            if (right - left + 1) - max_count > k:
                count[s[left]] -= 1
                left += 1
            
            # Update the maximum length of the valid window
            max_length = max(max_length, right - left + 1)
        
        return max_length

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        //BRUTE
        //REPLACE ALL CHARACTER HAVING LESS FREQ INSTED OF HAVING MORE FREQ
        //i.e. LEN-MAXFREQ
        // int maxlen=0;
        // for(int i=0;i<s.length();i++){
        //     int freq[]=new int[26];
        //     int maxfreq=0;
        //     for(int j=i;j<s.length();j++){
        //         freq[s.charAt(j)-'A']++;
        //         maxfreq=Math.max(maxfreq,freq[s.charAt(j)-'A']);
        //         int changes=j-i+1-maxfreq;
        //         if(changes<=k){
        //             maxlen=Math.max(maxlen,j-i+1);
        //         }
        //         else{
        //             break;
        //         }
        //     }
        // }
        // return maxlen;

        int freq[]=new int[26];
        int left=0;
        int maxfreq=0;
        int maxwid=0;
        for(int right=0;right<s.length();right++){
            freq[s.charAt(right)-'A']++;
            maxfreq=Math.max(maxfreq,freq[s.charAt(right)-'A']);
            int windowSize=right-left+1;
            if(windowSize-maxfreq>k){
                freq[s.charAt(left)-'A']--;
                left++;
            }
            windowSize=right-left+1;
            maxwid=Math.max(maxwid,windowSize);
        }
        return maxwid;
    }
}